src/webgpu/shader/validation/expression/call/builtin/textureSample.spec.ts

```diff
@@ -0,0 +1,223 @@
+      .combine('arrayIndexType', keysOf(kValuesTypes))
+      .beginSubcases()
+      .combine('value', [-9, -8, 0, 7, 8])
+  )
+  .fn(t => {
+    const { textureType, arrayIndexType, value } = t.params;
+    const arrayIndexArgType = kValuesTypes[arrayIndexType];
+    const args = [arrayIndexArgType.create(value)];
+    const { coords, offset = undefined } = kValidTextureSampleParameterTypes[textureType];
+
+    const coordWGSL = coords.create(0).wgsl();
+    const arrayWGSL = args.map(arg => arg.wgsl()).join(', ');
+    const offsetWGSL = offset ? `, ${offset.create(0).wgsl()}` : '';
+
+    const code = `
+@group(0) @binding(0) var s: sampler;
+@group(0) @binding(1) var t: ${textureType};
+@fragment fn fs() -> @location(0) vec4f {
+  let v = textureSample(t, s, ${coordWGSL}, ${arrayWGSL}${offsetWGSL});
+  return vec4f(0);
+}
+`;
+
+    const expectSuccess =
+      isConvertible(arrayIndexArgType, Type.i32) || isConvertible(arrayIndexArgType, Type.u32);
+    t.expectCompileResult(expectSuccess, code);
+  });
+
+g.test('offset_argument')
+  .desc(
+    `
+Validates that only incorrect offset arguments are rejected by ${builtin}
+`
+  )
+  .params(u =>
+    u
+      .combine('textureType', kTextureTypes)
+      // filter out types with no offset
+      .filter(
+        ({ textureType }) => kValidTextureSampleParameterTypes[textureType].offset !== undefined
+      )
+      .combine('offsetType', keysOf(kValuesTypes))
+      .beginSubcases()
+      .combine('value', [-9, -8, 0, 7, 8])
+  )
+  .fn(t => {
+    const { textureType, offsetType, value } = t.params;
+    const offsetArgType = kValuesTypes[offsetType];
+    const args = [offsetArgType.create(value)];
+    const {
+      coords,
+      array_index,
+      offset: offsetRequiredType = undefined,
+    } = kValidTextureSampleParameterTypes[textureType];
+
+    const coordWGSL = coords.create(0).wgsl();
+    const arrayWGSL = array_index ? ', 0' : '';
+    const offsetWGSL = args.map(arg => arg.wgsl()).join(', ');
+
+    const code = `
+@group(0) @binding(0) var s: sampler;
+@group(0) @binding(1) var t: ${textureType};
+@fragment fn fs() -> @location(0) vec4f {
+  let v = textureSample(t, s, ${coordWGSL}${arrayWGSL}, ${offsetWGSL});
+  return vec4f(0);
+}
+`;
+
+    const expectSuccess =
+      isConvertible(offsetArgType, offsetRequiredType!) && value >= -8 && value <= 7;
+    t.expectCompileResult(expectSuccess, code);
+  });
+
+g.test('offset_argument,non_const')
+  .desc(
+    `
+Validates that only non-const offset arguments are rejected by ${builtin}
+`
+  )
+  .params(u =>
+    u
+      .combine('textureType', kTextureTypes)
+      .combine('varType', ['c', 'u', 'l'])
+      // filter out types with no offset
+      .filter(
+        ({ textureType }) => kValidTextureSampleParameterTypes[textureType].offset !== undefined
+      )
+  )
+  .fn(t => {
+    const { textureType, varType } = t.params;
+    const {
+      coords,
+      array_index,
+      offset = Type.vec2f,
```

Comment:
Little bit odd to my not-so-familiar-with-JS-eye to have an assignment in here, then another assignment outside. Might be better to call this `offsetType` too, to match the other types.

Reply:
I'm conflicted here and would love your input. I picked those names originally because they match the spec

https://gpuweb.github.io/gpuweb/wgsl/#texturesample

But I'm happy to change them. They do seem ... confusing.  Thoughts?

The assignment here offset can be undefined since the set of textureSampler overloads, some include offset and some don't. So, even though all the textureSampler overloads with no offset are filtered out of the parameters, TypeScript complains later that offset might be undefined. I can probably just add a `!` at the right place instead of this `= Type.vec2f`

Reply:
Ah - I didn't even realize what this was doing. `offset = Type.vec2f` as a default (which is an invalid type), which gets replaced with `Type.vec2i` if its present on the `kValidTextureSampleParameterTypes` entry - which it always is? What a strange language.

I was wondering if you can get TypeScript to infer that `offset` is never undefined after filtering, but after 20 mins of playing around, I've conceded that my TS-foo is weak. Maybe @kainino0x might know the magic required to narrow the types here.

I'm happy to land this, but a comment here would be good to explain what this is doing.

Reply:
I removed it and added a `!` suppression later.